asterius/rts/rts.scheduler.mjs

```diff
@@ -0,0 +1,473 @@
+    return this.tsos.get(tid).addr;
+  }
+
+  getTSOret(tid) {
+    return this.tsos.get(tid).ret;
+  }
+
+  getTSOrstat(tid) {
+    return this.tsos.get(tid).rstat;
+  }
+
+  setTSOaddr(tid, addr) {
+    this.tsos.get(tid).addr = addr;
+  }
+
+  setTSOret(tid, ret) {
+    this.tsos.get(tid).ret = ret;
+  }
+
+  setTSOrstat(tid, rstat) {
+    this.tsos.get(tid).rstat = rstat;
+  }
+
+  getTSOid(tso) {
+    return this.memory.i32Load(tso + rtsConstants.offset_StgTSO_id);
+  }
+
+  /**
+   * Called from a generated safe FFI import call.
+   *
+   * @param ffiPromise Promise executing the FFI import code asynchronously.
+   */
+  returnFFIPromise(tid, ffiPromise) {
+    const tso_info = this.tsos.get(tid);
+    //console.log(`Thread ${tid}: return FFI Promise`);
+    tso_info.blockingPromise = ffiPromise;
+  }
+
+  /**
+   * Called when a thread stops for some reason.
+   */
+  returnedFromTSO(tid) {
+    const tso_info = this.tsos.get(tid);
+    const tso = tso_info.addr;
+    const reason = Number(this.memory.i64Load(
+                            this.symbolTable.MainCapability +
+                              rtsConstants.offset_Capability_r +
+                              rtsConstants.offset_StgRegTable_rRet
+                          ));
+
+    switch(reason) {
+      case 1: { // HeapOverflow
+
+        this.gc.performGC();
+
+        // put the thread back into the run-queue
+        // TODO: we should put it in front if it hasn't exceeded its time splice
+        this.runQueue.push(tid);
+        this.submitCmdWakeUp();
+        break;
+      }
+      case 2: { // StackOverflow
+        throw new WebAssembly.RuntimeError("StackOverflow");
+        break;
+      }
+      case 3: { // ThreadYielding
+        // put the thread back into the run-queue
+        this.runQueue.push(tid);
+        this.submitCmdWakeUp();
+        break;
+      }
+      case 4: { // ThreadBlocked
+        this.blockedTSOs.add(tid);
+
+        const why_blocked = Number(this.memory.i16Load(tso + rtsConstants.offset_StgTSO_why_blocked));
+
+        switch (why_blocked) {
+          case Blocked.OnCCall:
+          case Blocked.OnCCall_Interruptible: {
+            //console.log(`Thread ${tid}: blocked on FFI`);
+            // Wait for the FFI blocking promise and then requeue the TSO
+            tso_info.blockingPromise.then
+              (v => {
+                    //console.log(`Thread ${tid}: unblocked`);
+                    const [retTyp,retVal] = v;
+                    tso_info.ffiRet       = retVal;
+                    tso_info.ffiRetType   = retTyp;
+                    // Store return block symbol
+                    tso_info.ffiAsyncFunc = this.memory.i64Load(tso + rtsConstants.offset_StgTSO_ffi_func);
+
+                    this.blockedTSOs.delete(tid);
+                    this.runQueue.push(tid);
+                    this.submitCmdWakeUp();
+                  }
+              , e => {
+                    tso_info.ffiRetErr = e;
```

Comment:
Why don't we re-enqueue the thread and submit a wakeup command in case of an FFI `Promise` rejection here?

Reply:
Nice catch!